### PR TITLE
Initial guided proposal: multidisks first

### PR DESCRIPTION
**Pull request by dgdavid**
## Problem

Currently, the initial guided proposal tries first to install the product into individual disks, adjusting the default settings as much as possible (disabling snapshots, separated home, etc) before attempt to use several disks to fulfill the requirements. This is not useful in certain scenarios in which the product has been designed to be expanded over as many disks as possible, like SUSE Manager.

- https://trello.com/c/tSrlEnnb/
- Related to https://github.com/yast/yast-storage-ng/pull/936

## Solution

Add a new configuration value to the control file and use it to decide how many groups of candidate devices will be generated.

It has been called `multidisk_first` and it's `false` by default. When set to `true`, the initial proposal will be tried using only one group containing all available candidate devices, resulting in a proposal which will use several disks before adjusting the settings.

## Testing

- Added unit tests
- Also tested manually via _dud_.

## Screenshots

<details>
<summary>Click to show/hide screenshots</summary>

---

Screenshot_test_2019-08-11_01:09:00

<p align="center"><sup>Installing SUSE Manager Server</sup></p>

---

Screenshot_test_2019-08-11_01:30:22

<p align="center"><sup>Initial proposal uses as much devices as possible (1/2).</sup></p>

Screenshot_test_2019-08-11_01:30:30

<p align="center"><sup>Initial proposal uses as much devices as possible (2/2).</sup></p>

---

Screenshot_test_2019-08-11_01:31:07

<p align="center"><sup>Using the expert partitioner to see the initial proposal.</sup></p>

---

Screenshot_test_2019-08-11_01:32:22

<p align="center"><sup>Device graph of the initial proposal.</sup></p>

</details>

### CODE REVIEW

File: test/y2storage/initial_guided_proposal_test.rb [hunk 1]
```diff
@@ -193,6 +194,353 @@
         end
       end
 
+      context "and the proposal is set to use a multidisk_first approach" do
```
Comment (dgdavid) [inline test/y2storage/initial_guided_proposal_test.rb]: Thought: maybe it's time to split this long test file for the sake of legibility. Having this kind of "special" scenarios in a separated file looks good to me.

File: doc/old_and_new_proposal.md [hunk 2]
```diff
@@ -660,6 +661,10 @@ following options.
     consequence, the interface presented to the user by default will be
     different (containing different questions) depending of the value of this
     property.
+  * `multidisk_first`
+    When set to `true`, the initial proposal will be performed over all
+    available candidate disks, which means that the proposal will try to use
+    several disks before jumping to adjust the settings.
```
Comment (joseivanlopez) [inline doc/old_and_new_proposal.md]: NP: not sure if this version is better understandable:

When set to `true`, the initial proposal considers all disks as possible destination for the volumes to allocate. If set to `false`, the proposal tries to use individual disks to allocate the volumes, and it only uses all disks together when the volumes cannot be allocated in any individual disk.

File: src/lib/y2storage/proposal_settings.rb [hunk 3]
```diff
@@ -58,6 +58,14 @@ class ProposalSettings # rubocop:disable ClassLength
     # @return [:auto, :device] :auto by default
     attr_accessor :allocate_volume_mode
 
+    # Whether the initial proposal should use a multidisk approach first
+    #
+    # @note :ng format
+    #
+    # @return [Boolean] if true, the initial proposal will be tried using all
+    #   available candidate disk before jumping to a single disk approach.
```
Comment (joseivanlopez) [inline src/lib/y2storage/proposal_settings.rb]: Actually, single disk is not tried at all. Please, adapt the documentation.

File: src/lib/y2storage/proposal_settings.rb [hunk 4]
```diff
@@ -525,6 +535,7 @@ def apply_legacy_defaults
       self.other_delete_mode         ||= :ondemand
       self.root_space_percent        ||= 40
       self.allocate_volume_mode      ||= :auto
+      self.multidisk_first           ||= false
```
Comment (joseivanlopez) [inline src/lib/y2storage/proposal_settings.rb]: NP: for legacy is not needed to initialize this attribute.

File: src/lib/y2storage/proposal_settings.rb [hunk 5]
```diff
@@ -644,6 +655,7 @@ def ng_string_representation
       "    delete_resize_configurable: #{delete_resize_configurable}\n" \
       "    lvm_vg_strategy: #{lvm_vg_strategy}\n" \
       "    lvm_vg_size: #{lvm_vg_size}\n" \
+      "    multidisk_first: #{multidisk_first}\n" \
```
Comment (joseivanlopez) [inline src/lib/y2storage/proposal_settings.rb]: NP: maybe we should start thinking about using pretty print instead of manually fomatting the output.

File: test/y2storage/initial_guided_proposal_multidisk_first_test.rb [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+            include_examples "make a proposal using as much devices as possible"
+          end
+
+          context "without separated volume groups" do
+            let(:separate_vgs) { false }
+
+            it "spreads the proposal as much as possible" do
+              proposal.propose
+
+              expect(used_devices).to contain_exactly("/dev/sda", "/dev/sdb", "/dev/sdd")
```
Comment (joseivanlopez) [inline test/y2storage/initial_guided_proposal_multidisk_first_test.rb]: NP: I would also check that no volume groups are created.

File: test/y2storage/initial_guided_proposal_multidisk_first_test.rb [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+        context "using LVM" do
+          let(:lvm) { true }
+
+          context "with separated volume groups" do
+            let(:separate_vgs) { true }
+
+            it "spreads the proposal as much as possible" do
+              proposal.propose
+
+              expect(used_devices).to contain_exactly("/dev/sda", "/dev/sdb", "/dev/sdd")
```
Comment (joseivanlopez) [inline test/y2storage/initial_guided_proposal_multidisk_first_test.rb]: Umm, could you try it with a very very big sda? I would expect it only uses the necessary devices.

File: test/y2storage/initial_guided_proposal_multidisk_first_test.rb [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+          context "with separated volume groups" do
+            let(:separate_vgs) { true }
+
+            context "and there is room for the full proposal in a single device" do
+              let(:sdb_size) { 500.GiB }
+
+              it "makes the full proposal using a single device" do
+                proposal.propose
+
+                expect(used_devices).to contain_exactly("/dev/sdb")
```
Comment (joseivanlopez) [inline test/y2storage/initial_guided_proposal_multidisk_first_test.rb]: Yes, I expect the same with `allocate_mode: :auto, lvm: true, separate_vgs: true`.

File: test/y2storage/initial_guided_proposal_multidisk_first_test.rb [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+
+            # NOTE: with current settings, a minimal proposal fits in sda, sdb or sdd
+            context "and there is room for a minimal proposal in a single device" do
+              it "makes the full proposal" do
+                proposal.propose
+
+                expect(proposed_vg_names).to include("vg-foobar")
+              end
+
+              it "uses multiple disks" do
```
Comment (joseivanlopez) [inline test/y2storage/initial_guided_proposal_multidisk_first_test.rb]: Why does it use multiple disks?

File: test/y2storage/initial_guided_proposal_multidisk_first_test.rb [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+              it "makes the minimal proposal" do
+                proposal.propose
+
+                expect(proposed_vg_names).to_not include("vg-foobar")
+              end
+
+              it "uses multiple disks" do
+                proposal.propose
+
+                expect(used_devices.count).to be > 1
```
Comment (joseivanlopez) [inline test/y2storage/initial_guided_proposal_multidisk_first_test.rb]: I would prefer to check which disks are used exactly.

File: doc/old_and_new_proposal.md (hunk 2 above)
Comment (dgdavid) [reply doc/old_and_new_proposal.md]: Thanks!

File: src/lib/y2storage/proposal_settings.rb (hunk 5 above)
Comment (dgdavid) [reply src/lib/y2storage/proposal_settings.rb]: If I remember correctly, it was discussed during the planning at it will be addressed in the ["Better logging of proposal settings"](https://trello.com/c/tYqIvWA4/1211-1-sle15-sp1-storage-ng-better-logging-of-proposal-settings) PBI.

Comment (joseivanlopez) [reply src/lib/y2storage/proposal_settings.rb]: Oh, right! Thanks! As you see, I have bad memory.

File: src/lib/y2storage/initial_guided_proposal.rb [hunk 11]
```diff
@@ -246,10 +263,20 @@ def groups_of_candidate_devices
...
+
+      # FIXME: the use of .reverse method changes the expected precedence
+      # (alphabetically) when the "contiguous" devices has the equal sizes. See
+      # the example:
+      #
+      #   disk_names.map { |n| initial_devicegraph.find_by_name(n) }.compact.sort_by(&:size)
+      #   [<Disk /dev/sdc 20 GiB>, <Disk /dev/sdd 20 GiB>, <Disk /dev/sdb 80 GiB>]
+      #
+      #   disk_names.map { |n| initial_devicegraph.find_by_name(n) }.compact.sort_by(&:size).reverse
+      #   [<Disk /dev/sdb 80 GiB>, <Disk /dev/sdd 20 GiB>, <Disk /dev/sdc 20 GiB>]
```
Comment (joseivanlopez) [inline src/lib/y2storage/initial_guided_proposal.rb]: NP: I would mention the expected result:

```
# Expected result:
#
#   [<Disk /dev/sdb 80 GiB>, <Disk /dev/sdc 20 GiB>, <Disk /dev/sdd 20 GiB>]
```

File: src/lib/y2storage/proposal_settings.rb (hunk 5 above)
Comment (dgdavid) [reply src/lib/y2storage/proposal_settings.rb]: Don't worry. Mine is worst :cry:

File: test/y2storage/initial_guided_proposal_multidisk_first_test.rb [hunk 12]
```diff
@@ -0,0 +1,526 @@
...
+                  .to contain_exactly("/dev/sda", "/dev/sdb", "/dev/sdc", "/dev/sdd")
+              end
+            end
+          end
+        end
+
+        context "without separated volume groups" do
+          let(:separate_vgs) { false }
+
+          context "and there is room for the full proposal" do
```
Comment (joseivanlopez) [inline test/y2storage/initial_guided_proposal_multidisk_first_test.rb]: This external context is leftover.

File: test/y2storage/initial_guided_proposal_multidisk_first_test.rb [hunk 13]
```diff
@@ -0,0 +1,526 @@
...
+
+              it "makes the proposal in a single device" do
+                proposal.propose
+
+                expect(used_devices).to contain_exactly("/dev/sdb")
+              end
+            end
+
+            context "and the proposal does not fit in a single device" do
+              # NOTE: there is needed to create three partitions of at least 10GiB plus the boot
```
Comment (joseivanlopez) [inline test/y2storage/initial_guided_proposal_multidisk_first_test.rb]: NP: s/there is/it is/

File: test/y2storage/initial_guided_proposal_multidisk_first_test.rb [hunk 14]
```diff
@@ -0,0 +1,526 @@
...
+              # The minimal proposal including all volumes needs 30GiB
+              let(:sda_size) { 8.GiB }
+              let(:sdb_size) { 8.GiB }
+              let(:sdc_size) { 10.GiB }
+              let(:sdd_size) { 11.GiB }
+
+              # FIXME: it seems that SpaceMaker is having troubles to assign
+              # space at the time to create multiple LVM VGs when
+              # allocate_volume_mode => :auto and separate_vgs => true
+              xit "creates as partition all volumes without separate_vg_name" do
```
Comment (joseivanlopez) [inline test/y2storage/initial_guided_proposal_multidisk_first_test.rb]: NP: I would reword it as: "creates partitions for all volumes without separate_vg_name".

File: test/y2storage/initial_guided_proposal_multidisk_first_test.rb [hunk 15]
```diff
@@ -0,0 +1,526 @@
...
+                proposal.propose
+
+                expect(created_mount_paths).to include("/")
+              end
+
+              # FIXME: it seems that SpaceMaker is having troubles to assign
+              # space at the time to create multiple LVM VGs when
+              # allocate_volume_mode => :auto and separate_vgs => true
+              xit "creates as LVM VG all volumes with separate_vg_name" do
```
Comment (joseivanlopez) [inline test/y2storage/initial_guided_proposal_multidisk_first_test.rb]: NP: I would reword it as: "creates LVM VGs for all volumes with separate_vg_name".
